File: 92. Reverse Linked List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        int chg_len = n-m+1;
        ListNode *pre_head = NULL;
        ListNode *result   = head;
        while(--m){
            pre_head = head;
            head = head->next;
        }
        ListNode *new_tail = head;
        ListNode *new_head = NULL;
        while(head && chg_len){
            ListNode *next = head->next;
            head->next = new_head;
            new_head = head;
            head = next;
            chg_len--;
        }
        new_tail->next = head;
        if(pre_head){
            pre_head->next = new_head;
        }else{
            result = new_head;
        }
        return result;
    }
};

File: 92. Reverse Linked List II.py
class Solution:
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        chg_len = int(n-m+1)
        
        pre_head = None
        result = head
        while m>1:
            m-=1
            pre_head = head
            head = head.next
            
        new_tail = head
        new_head = None
        
        while head and chg_len :
            new_next = head.next
            head.next = new_head
            new_head = head
            head = new_next
            chg_len-=1
        new_tail.next = head
        if pre_head:
            pre_head.next = new_head
        else:
            result = new_head
        return result
